Add Typesense

ruslandoga opened the pull request:
This PR integrates Typesense into Hexdocs.

**TODOs:**
- [x] should Broadway task fail on failed indexing, or should it just log an error?
   __Logging excessively for now__
- [x] should indexing happen in the same step as upload to object storage or can it be parallel?
   __Indexing in the same step for now__
- [x] index `proglang`
- [x] index Erlang, what's its search_data equivalent?
   __The format is the same:__ https://github.com/hexpm/hexdocs/pull/44#issuecomment-2381604920
- [x] index Gleam: https://github.com/hexpm/hexdocs/pull/44/commits/ce7af3bb0256467d7db048ddb2b3a10cb66fb3b6
  __Commit reverted, approach needs discussion with Gleam team__
- [x] more tests

CI results: https://github.com/ruslandoga/hexdocs/pull/1

ruslandoga committed:
fix ci

ruslandoga commented on lib/hexdocs/search.ex:
```diff
@@ -0,0 +1,80 @@
+
+    case :hackney.post(url, headers, ndjson) do
+      {:ok, 200, _resp_headers, ref} ->
+        process_results(package, version, ref)
```
Typesense responds with lines of "status" jsons for each indexed document: https://typesense.org/docs/26.0/api/documents.html#index-multiple-documents

```json
{"success": true}
{"success": false, "error": "Bad JSON.", "document": "[bad doc]"}
```

This function just logs an error per each when .success==false.

ruslandoga commented on test/hexdocs/queue_test.exs:
```diff
@@ -576,4 +644,38 @@ defmodule Hexdocs.QueueTest do
+        {"name": "type", "type": "string", "facet": true},
+        {"name": "title", "type": "string"},
+        {"name": "doc", "type": "string"},
+        {"name": "package", "type": "string", "facet": true}
```
This is the current collection schema.

ruslandoga commented on lib/hexdocs/queue.ex:
```diff
@@ -110,6 +110,12 @@ defmodule Hexdocs.Queue do
 
         case Hexdocs.Tar.unpack(body, repository: repository, package: package, version: version) do
           {:ok, files} ->
+            if repository == "hexpm" do
```
I assume only public packages would be indexed.

ruslandoga commented on lib/hexdocs/search.ex:
```diff
@@ -0,0 +1,80 @@
+    url = url("collections/#{@collection}/documents/import?action=create")
+    headers = headers([{"content-type", "text/plain"}])
+
+    case :hackney.post(url, headers, ndjson) do
```
Since it's possible to use real Typesense in tests, I'm not hiding it behind a behavior and make "real" HTTP requests here. I am totally OK to change it and make it similar to Hexdocs.Store behaviors if it's easier for the maintainers.

ericmj replied:
I think this is fine but we should put it behind a flag that is excluded by default and add a docker-compose file that starts typesense locally.

ericmj replied:
But if this is called from other code that is also tested then we need a "local" implementation that essentially no-ops.

You can see this pattern for the Fastly code: https://github.com/hexpm/hexdocs/tree/d7a408dc94ea5389481d007baff03e385329638b/lib/hexdocs/cdn.

ruslandoga replied:
> I think this is fine but we should put it behind a flag that is excluded by default 

This is done automatically right now if Typesense is not detected: https://github.com/hexpm/hexdocs/pull/44/files#diff-443233b9c679556d43885b3f3babee59b9c5c33da48887d591513e423d558f3bR7-R22

The warning also includes a Docker command to start Typesense. I'll try to make it appear only on `mix test --include typesense` instead of all the time.

> But if this is called from other code that is also tested

Right now no other code uses `search_data-....js` files in tests, and that is one of the prerequisites for Typesense indexing.

---

I'll be refactoring this PR tomorrow, and I'll try putting Hexdocs.Search behind a behaviour with Typesense being one of the implementations. And I'll add a compose file :)

josevalim commented:
Thank you @ruslandoga! I am currently on holidays but I will try to carve some time sooner than later to give you feedback. /cc @wojtekmach

josevalim commented:
>  index erlang, what's its search_data equivalent?

Erlang uses ExDoc now, so they have the exact same structure. However, we will need to either poll them or ask them to ping us once they publish a new version or ask them to push their docs to Hexdocs! I'd say we can postpone this to a follow up pull request.

On the other hand, I believe Gleam does not have the data in the format we need (we created our `search_data.js` somewhat recently thinking exactly about this). They would need to update their tools first, so we can postpone this conversation too.

wojtekmach commented:
@ruslandoga I'll try to review this sooner than later but I have a lot on my plate until October 15th, after that it's gonna be my number one priority to see this through. Sorry about that.

ericmj commented on lib/hexdocs/queue.ex:
```diff
@@ -250,4 +256,61 @@ defmodule Hexdocs.Queue do
+      end)
+
+    unless maybe_json do
+      Logger.error(
```
Since packages that weren't built with hexdocs won't have this.

```suggestion
      Logger.info(
```

ericmj commented on lib/hexdocs/queue.ex:
```diff
@@ -250,4 +256,61 @@ defmodule Hexdocs.Queue do
+        end
+      end
+
+    if maybe_docs do
```
We don't need this `if` if we check for `nil` inside the `case`.

ruslandoga committed:
Update lib/hexdocs/queue.ex

Co-authored-by: Eric Meadows-Jönsson <eric.meadows.jonsson@gmail.com>

ruslandoga committed:
add compose

ruslandoga committed:
hide typesense behind impl

ruslandoga committed:
simplify find_search_items/3

ruslandoga committed:
cleanup test

wojtekmach commented on test/test_helper.exs:
```diff
@@ -3,4 +3,25 @@ File.mkdir("tmp")
+      {:error, :econnrefused} -> false
+    end
+
+  unless typesense_available? do
```
awesome DX!

wojtekmach commented on lib/hexdocs/search/typesense.ex:
```diff
@@ -0,0 +1,76 @@
+      end)
+
+    url = url("collections/#{@collection}/documents/import?action=create")
+    headers = headers([{"content-type", "text/plain"}])
```
I was trying to change this content type value but no matter what I set, their API sends text/plain back as opposed to, ideally, application/x-ndjson. Was this the reason you put text/plain here? Any idea why they don't specify JSON-related content type in response?

ruslandoga replied:
I copied the content type from the curl example in the docs: https://typesense.org/docs/27.1/api/documents.html#index-multiple-documents -- I guess it only accepts ndjson so content-type=text/plain can be removed: https://github.com/hexpm/hexdocs/pull/44/commits/10a852c284d4f38cf72fbe1b7d33fc45916f0a36

And the response content-type is hardcoded: https://github.com/typesense/typesense/blob/e9cc6abfb2eeb0c464ab64f35b608acf196a7d21/src/core_api.cpp#L1404

wojtekmach replied:
Gotcha, thank you for digging this up.

wojtekmach commented on lib/hexdocs/search/typesense.ex:
```diff
@@ -0,0 +1,76 @@
+    headers = headers([{"content-type", "text/plain"}])
+
+    case :hackney.post(url, headers, ndjson) do
+      {:ok, 200, _resp_headers, ref} ->
```
My understanding is we're later reading the entire response body so we might as well do it in one go:

```suggestion
    case :hackney.post(url, headers, ndjson, [:with_body]) do
      {:ok, 200, _resp_headers, body} ->
```

that being said, can we use `Hexdocs.HTTP` for consistency? If it's missing some features for this use case and it's not too much hassle we'd appreciate adjusting it but I wouldn't sweat it too much, I'd like to eventually use Req here anyway.

ruslandoga replied:
We read the entire response but we don't persist it. I guess it's small enough to be read and accumulated in full. 

Done in https://github.com/hexpm/hexdocs/pull/44/commits/0c6f014df0a714e8d505eb551a96c0a1d538ce16

wojtekmach commented on lib/hexdocs/search/typesense.ex:
```diff
@@ -0,0 +1,76 @@
+        ndjson
+        |> String.splitter("\n")
+        |> Stream.each(fn json ->
+          case Jason.decode!(json) do
```
dunno if it'd make a difference performance wise but we should be able to use the new :json module for this:

```elixir
iex> binary = ~s({"x":1}\n{"x":2}\n{"x":3})
iex> acc = nil
iex> {decoded, acc, binary} = :json.decode(binary, acc, %{})
{%{"x" => 1}, nil, "{\"x\":2}\n{\"x\":3}\n"}
iex> {decoded, acc, binary} = :json.decode(binary, acc, %{})
{%{"x" => 2}, nil, "{\"x\":3}\n"}
iex> {decoded, acc, binary} = :json.decode(binary, acc, %{})
{%{"x" => 3}, nil, ""}
```

We can totally ship this as is, I am just mentioning this in case it piques your interest. Not gonna lie, I'd love :json-based _ndjson_ eager and streaming decoder for Req. ;)

ruslandoga replied:
Hexdocs is using Erlang 25 right now: https://github.com/hexpm/hexdocs/blob/2d7c25230b0a06bf395828fecca05b45858204dc/Dockerfile#L1

I can open a separate PR that updates the runtime to the latest versions and then rebase this PR and switch to `:json` once merged.

wojtekmach replied:
Yeah, that'd be great and appreciated, no rush though!

ruslandoga replied:
> no rush though!

Updating versions is my favorite activity for some reason: https://github.com/hexpm/hexdocs/pull/45 :)

wojtekmach commented on lib/hexdocs/search/typesense.ex:
```diff
@@ -0,0 +1,76 @@
+
+  @behaviour Hexdocs.Search
+
+  @collection "hexdocs"
```
random: I tried changing this string to "hexdocsv2" and I started getting 404s when indexing, is this hardcoded somewhere else?

ruslandoga replied:
Right. I made the collection name configurable in https://github.com/hexpm/hexdocs/pull/44/commits/6116a98a8b3c351bdd4e46c44bdba76863a4a31e, it used to be hardcoded in tests.

wojtekmach reviewed (commented):
This is very exiting! I left some comments below.

ruslandoga committed:
include typesense in ci tests

ruslandoga committed:
add post in hexdocs.http

ruslandoga committed:
read typesense indexing response eagerly

ruslandoga committed:
make collection configurable

ruslandoga committed:
remove content-type: text/plain from import

ruslandoga committed:
update typesense to 27.1

ruslandoga committed:
do typesense indexing last

ruslandoga committed:
extract indexing to separate function, wrap in log lines

josevalim commented on lib/hexdocs/queue.ex:
```diff
@@ -250,4 +266,96 @@ defmodule Hexdocs.Queue do
+        extract_elixir_search_items(package, version, content)
+
+      {:gleam, content} ->
+        extract_gleam_search_items(package, version, content)
```
If we are going to index Gleam as is, I'd like explicit confirmation from them that they are willing to make search-data.js "public" and retain backwards compatibility. We should also double check that their data is ready for indexing. Previous ExDoc versions had a stripped down version of the docs (we would convert markdown to html and then strip the html) which sometimes would lead to suboptimal results. :)

wojtekmach replied:
Maybe it is best if Gleam follows the same format as ExDoc, so we don't have to support multiple different formats over time? The one in ExDoc has been made generic over time exactly for cases like this.

josevalim replied:
That's not a bad idea. I just pushed a proglang field to ExDoc, so they could adhere to that and set the proglang as "gleam".

ruslandoga replied:
I think I'll revert the index-gleam commit, and return it once we either have a confirmation from the Gleam team that the format of their search_data is stable or they switch to the ExDoc format. I also think the latter would be best since the current format doesn't map one to one to the chosen Typesense schema which is based on ExDoc. For example, there is no `type` field (like module, function, etc.) in Gleam, which might be important if we might need to use it for ranking. Also titles for types and functions don't include module names, and that affects the search behavior.

josevalim commented on lib/hexdocs/queue.ex:
```diff
@@ -250,4 +266,96 @@ defmodule Hexdocs.Queue do
+    "window.Gleam.initSearch(" <> search_items_js = search_data
+
+    search_items =
+      case Jason.decode(search_items_js) do
```
Should we strip the trailing parens before trying to decode? Or is it not so trivial?

ruslandoga replied:
> Should we strip the trailing parens before trying to decode?

Yes, that's how it should've been done :)

josevalim commented on lib/hexdocs/queue.ex:
```diff
@@ -250,4 +266,96 @@ defmodule Hexdocs.Queue do
+
+    case search_data do
+      {:elixir, content} ->
+        extract_elixir_search_items(package, version, content)
```
Perhaps it would be more precise to say: `extract_ex_doc_search_items`. I will make a comment on getting the proglang a bit down below.

josevalim commented on lib/hexdocs/queue.ex:
```diff
@@ -250,4 +266,96 @@ defmodule Hexdocs.Queue do
+
+    case search_data do
+      %{"items" => items} ->
+        items
```
I have added the proglang to ExDoc just now. It means it won't be available for old ones, but we can get it like this:

```elixir
language =
  case search_data do
    %{"proglang" => proglang} =>
      proglang

    %{"items" => items} =>
      # For older versions, we infer the module
      if Enum.any?(items, &elixir_module?/1), do: "elixir", else: "erlang"
  end
```

where

```elixir
defp elixir_module?(%{type: "module", title: <<first_letter, _::binary>>}) when first_letter in ?A..?Z, do: true
defp elixir_module?(_), do: false
```

ruslandoga committed:
refactor

ruslandoga committed:
handle more errors

ruslandoga committed:
add proglang to collection

ruslandoga committed:
refactor tests a bit

ruslandoga committed:
test find_search_items/3

ruslandoga committed:
test package delete

ruslandoga committed:
refactor log tests

ruslandoga committed:
add bad document test

ruslandoga commented:
👋 

I think it's ready for review now :)

Updates since the last review: 
- removed Gleam indexing (for now)
- added `proglang` field to the schema
- moved some code around in attempt to make it cleaner
- added more tests

ruslandoga committed:
remove search_data_js format error logs when there is no search_data_js

ericmj commented on lib/hexdocs/search/search.ex:
```diff
@@ -0,0 +1,94 @@
+    search_data_js =
+      Enum.find_value(files, fn {path, content} ->
+        case Path.basename(path) do
+          "search_data-" <> _digest -> content
```
We cannot trust this data since it's user provided, can they do anything dangerous by providing something we don't expect? Maybe we should do some rudimentary validation?

ruslandoga replied:
They can provide long strings like https://github.com/cloudpods-dev/docker-engine-api-elixir/blob/813cc557da483f623a8f484db04efc7e58db0376/lib/docker_engine_api/api/container.ex#L67, but Typesense seems to handle it fine. We can check for content size, maybe. I think if Typesense doesn't like the payload, it would simply reject it.

ruslandoga replied:
I added a test that checks that invalid fields in search items (like `type` being a map instead of a string, or `doc` being a list) are rejected: https://github.com/hexpm/hexdocs/pull/44/commits/8d58e4fbce076743d7b00e27ff595934a5d4360c

ruslandoga committed:
add typesense

ruslandoga committed:
rm releases.exs

ruslandoga committed:
read Typesense collection from env

ruslandoga committed:
switch to :json

ruslandoga committed:
update tests

ruslandoga committed:
add tests for invalid search items

ruslandoga commented on config/runtime.exs:
```diff
@@ -5,6 +5,9 @@ if config_env() == :prod do
     hexpm_secret: System.fetch_env!("HEXDOCS_HEXPM_SECRET"),
+    typesense_url: System.fetch_env!("TYPESENSE_URL"),
+    typesense_api_key: System.fetch_env!("TYPESENSE_API_KEY"),
+    typesense_collection: System.fetch_env!("TYPESENSE_COLLECTION"),
```
I created `hexdocs-test` collection on Typesense Cloud, https://cloud.typesense.org/clusters/ent97o5sv4dzx2f0p/collections

I think we can use it during alpha/beta testing.

ruslandoga commented on config/runtime.exs:
```diff
@@ -5,6 +5,9 @@ if config_env() == :prod do
     hexpm_url: System.fetch_env!("HEXDOCS_HEXPM_URL"),
     hexpm_secret: System.fetch_env!("HEXDOCS_HEXPM_SECRET"),
+    typesense_url: System.fetch_env!("TYPESENSE_URL"),
+    typesense_api_key: System.fetch_env!("TYPESENSE_API_KEY"),
```
This would be the "Admin API Key" for https://cloud.typesense.org/clusters/ent97o5sv4dzx2f0p cluster, it can be downloaded from the dashboard.

ruslandoga commented on config/runtime.exs:
```diff
@@ -5,6 +5,9 @@ if config_env() == :prod do
     port: System.fetch_env!("HEXDOCS_PORT"),
     hexpm_url: System.fetch_env!("HEXDOCS_HEXPM_URL"),
     hexpm_secret: System.fetch_env!("HEXDOCS_HEXPM_SECRET"),
+    typesense_url: System.fetch_env!("TYPESENSE_URL"),
```
And this would be `https://ent97o5sv4dzx2f0p.a1.typesense.net`

ruslandoga commented:
👋 

Just wanted to check if there’s anything else we'd need to address before it's merged?

wojtekmach reviewed (approved):
Looks good to me. We plan to test this out in our staging server before merging. Thank you for all the work!

wojtekmach committed:
Merge branch 'main' into add-typesense

wojtekmach committed:
Prepare hexdocs staging for typesense

wojtekmach committed:
wip

wojtekmach committed:
Revert "wip"

This reverts commit 26169bbd03161fe5d3b3ca629afa2f8b2aa95132.

wojtekmach commented:
@ruslandoga this is now deployed to https://staging.hexdocs.pm and I have published a test package and it was correctly indexed into hexdocs-test collection.

Is this ready to publish to prod? What are the next steps?

Thank you for all of your work on this and apologies for delays.

ruslandoga commented:
👋 @wojtekmach 

Yes, I think it's ready for prod! I think the next step would be integrating global search into ex_doc. I can open a PR!

ruslandoga commented:
We can create a new collection if needed or we can continue using `hexdocs-test`

I think it's possible to clone / fork collections in Typesense cloud, and if not, it's pretty easy to move the data around, since I guess by the nature of Hexdocs, the indexed data is immutable. So it should be OK either way.

wojtekmach commented:
Got it, excellent!

Could you create hexdocs-staging and hexdocs-prod collections for us? We tend to follow that particular convention for external services.

Should we backfill some data? Latest versions of all packages? All versions of all packages? We can also do nothing for now and make a decision when all pieces are ready. cc @josevalim

ruslandoga commented:
I've created `hexdocs-staging` and `hexdocs-prod` collections with the same schema.

Regarding backfilling, it definitely helps during development but I tend to use a local Typesense instance. It's a bit outdated by now (e.g. missing proglang), but here're some docs I used before: https://hexdocs-artifacts.s3.eu-central-003.backblazeb2.com/docs_from_tarballs_all_versions.jsonl.zst (it's 190MB compressed, 4.4G uncompressed)

```sh
$ curl https://hexdocs-artifacts.s3.eu-central-003.backblazeb2.com/docs_from_tarballs_all_versions.jsonl.zst -O
$ zstd docs_from_tarballs_all_versions.jsonl.zst -d
# add proglang=elixir to all entries
$ jq '. + {proglang: "elixir"}' docs_from_tarballs_all_versions.jsonl > docs.jsonl
$ docker compose up typesense -d

# https://typesense.org/docs/27.1/api/collections.html#with-pre-defined-schema
$ curl "http://localhost:8108/collections" \
       -X POST \
       -H "Content-Type: application/json" \
       -H "X-TYPESENSE-API-KEY: hexdocs" \
       -d '{"fields": [
    {"facet": true, "name": "proglang", "type": "string"},
    {"facet": true, "name": "type", "type": "string"},
    {"name": "title", "type": "string"},
    {"name": "doc", "type": "string"},
    {"facet": true, "name": "package", "type": "string"}
  ],
  "name": "hexdocs-local",
  "token_separators": [".", "_", "-", " ", ":", "@", "/"]
}'

# https://typesense.org/docs/27.1/api/documents.html#import-a-jsonl-file
$ curl "http://localhost:8108/collections/hexdocs-local/documents/import?action=create" \
       -X POST \
       -T docs.jsonl \
       -H "X-TYPESENSE-API-KEY: hexdocs"

# sanity check
$ curl -H "X-TYPESENSE-API-KEY: hexdocs" "http://localhost:8108/collections/hexdocs-local/documents/777777"
```

josevalim commented:
@ericmj @wojtekmach @ruslandoga let's definitely backfill. We only support recent ExDoc versions anyway, which will act as a filter.

A couple things to figure out:

1. Will Elixir releases be indexed automatically since we push to hexdocs.pm or does it require special treatment?
2. We should do something for Erlang in particular
3. Are we going to make typesense address public or are we going to expose it through hex.pm and use fast.ly to rewrite to typesense?

Overall, my next step suggestion is to build a home page for searching within a given set of packages. This thread on ElixirForum has a good example: https://elixirforum.com/t/hexdocs-search-engine-for-us-devs/46814/1

We could use it at least in two places:

1. A new hexdocs.pm home? We can store the selected packages in your cookies

2. `mix hex.search` which opens up a page with all of the hex packages of your lock file filled in (similar to `mix hex.outdated`). We can rename the existing `mix hex.search` to `mix hex.find`

For ExDoc, we would need to work on the related packages feature: https://github.com/elixir-lang/ex_doc/issues/1811

My idea is that we would be able to store this information as a .json file as well. So `phoenix_html` can say `related_deps: :phoenix` and we just query https://hexdocs.pm/phoenix/related_deps.json

josevalim commented:
I opened new issues: #46 #47 #48 #49.

josevalim commented:
This one looks good to merge to me. We can continue in the other issues.

wojtekmach commented:
Sounds good, I'll finish infrastructure setup and deploy this to prod soon.

wojtekmach commented:
This is now running on prod, so far so good!

josevalim commented:
Amazing work @ruslandoga !
